group/assignment5/main.cpp

```diff
@@ -0,0 +1,194 @@
+#include <iostream>
+#include <cmath>
+#include <vector>
+#include <unordered_map>
+#include "codeu_test_lib.h"
+
+using namespace std;
+
+vector<char> findAlphabetCharacters(const vector<string> &dictionary) {
+    vector<bool> charsInDict((unsigned int)pow(2, CHAR_BIT), false);
+    vector<char> alphabetChars;
+    for (auto word : dictionary) {
+        for (char c : word) {
+            if (!charsInDict[c]) {
+                charsInDict[c] = true;
+                alphabetChars.push_back(c);
+            }
+        }
+    }
+    return alphabetChars;
+}
+
+bool createCharPriorityGraph(vector<string> &dictionary, unordered_map<char, vector<bool>> &charPriorityGraph, unordered_map<char, int> &numOfPriorCharacters) {
+    while (dictionary.size() > 0 && dictionary[0].size() == 0) {
+        dictionary.erase(dictionary.begin());
+    }
+    if (dictionary.size() < 2) {
+        return true;
+    }
+    char prevChar = dictionary[0][0];
+    unordered_map<char, vector<string>> suffixesByCharacters;
+    suffixesByCharacters[prevChar] = vector<string>(0);
+    for (string word : dictionary) {
+        if (word.size() == 0) {
+            continue;
+        }
+        char currentChar = word[0];
+        if (currentChar != prevChar) {
+            if (suffixesByCharacters.find(currentChar) != suffixesByCharacters.end()) { // found a circle in characters - inconsistent dictionary
+                return false;
+            }
+            // mark prevChar as prior to currentChar and add new suffixes vector for current char
+            charPriorityGraph[prevChar][currentChar] = true;
+            suffixesByCharacters[currentChar] = vector<string>(0);
+            numOfPriorCharacters[currentChar]++;
+        }
+        word.erase(0, 1);
```

Comment:
Erase function effectively reduces the container size, but because vectors use an array as their underlying storage, erasing elements in positions other than the vector end causes the container to relocate all the elements after the segment erased to their new positions. This way solution is more cleaner, but it would be efficient if you just track current position in the word.

Reply:
Not only here but dictionary.erase(dictionary.begin()); this one too